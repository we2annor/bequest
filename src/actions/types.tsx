export const ADD_CONTACT = 'ADD_CONTACT';
export const FETCH_CONTACTS = 'FETCH_CONCTACTS';
export const FETCH_CONTACT = 'FETCH_CONTACT';
export const DELETE_CONTACT = 'DELETE_CONTACT';
export const EDIT_CONTACT = 'EDIT_CONTACT';

export interface AddressesState {
    addressBook: Address;
    //id?:ID
}

export interface Address {
    houseNo: string;
    streetName: string;
    locality: string;
    postTown: string;
    county: string;
    postCode: string;
    id: number
}

interface ID {
    id: number
}

export type AddressKey = keyof Address;

export const InitialAddressValues = {
    houseNo:'',
    streetName:'',
    locality: '',
    postTown : '',
    county: '',
    postCode: '',
    id: 0
}

export interface AddressFormProps {
  selectedAddress: Address;
  postcode: string;
  addContact: (formValues:Address)=>void;
}

export interface AddNewContact {
    type: typeof ADD_CONTACT
    payload: Address
}

export interface FetchContact {
    type: typeof FETCH_CONTACT;
    payload: Address
}

export interface FetchContacts {
    type: typeof FETCH_CONTACTS;
    payload: Address
}

export interface EditContact {
    type: typeof EDIT_CONTACT;
    payload: Address
}

export interface DeleteContact {
    type : typeof DELETE_CONTACT;
    payload: ID
}

export type AddressBookActionTypes = AddNewContact | FetchContact | FetchContacts | EditContact | DeleteContact ; 