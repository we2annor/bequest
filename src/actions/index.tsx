import { Dispatch } from "redux";
import {
  ADD_CONTACT,
  FETCH_CONTACT,
  FETCH_CONTACTS,
  EDIT_CONTACT,
  DELETE_CONTACT,
  Address,
} from "./types";
import { addressBook } from "../api/addressBook";
import history from "../components/history";

export const addContact = (formValues: Address) => async (
  dispatch: Dispatch
) => {
  const response = await addressBook.post("/addressBook", formValues);
  dispatch({ type: ADD_CONTACT, payload: response.data });
  console.log("add contact", formValues);
  history.push("/");
};

export const fetchContact = (id: number) => async (dispatch: Dispatch) => {
  const response = await addressBook.get(`/addressBook/${id}`);
  dispatch({
    type: FETCH_CONTACT,
    payload: response.data,
  });
};

export const editContact = (id: number) => async (dispatch: Dispatch) => {
  const response = await addressBook.get(`/addressBook/${id}`);
  dispatch({
    type: EDIT_CONTACT,
    payload: response.data,
  });
};

export const deleteContact = (id: any) => async (dispatch: Dispatch) => {
  await addressBook.get(`/addressBook/${id}`);
  dispatch({ type: DELETE_CONTACT, payload: id });
};

export const fetchContacts = () => async (dispatch: Dispatch) => {
  const response = await addressBook.get("/addressBook");
  dispatch({
    type: FETCH_CONTACTS,
    payload: response.data,
  });
};
