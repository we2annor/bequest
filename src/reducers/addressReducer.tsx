import _ from "lodash";
import {
  ADD_CONTACT,
  FETCH_CONTACT,
  FETCH_CONTACTS,
  EDIT_CONTACT,
  //DELETE_CONTACT,
  AddressesState,
  AddressBookActionTypes,
} from "../actions/types";

const initialState : AddressesState = Object({})

export const addressReducer = (
  state = initialState,
  action: AddressBookActionTypes
): AddressesState => {
  switch (action.type) {
    case FETCH_CONTACT:
      return { ...state, [action.payload.id]: action.payload };
    case FETCH_CONTACTS:
      return { ...state, ..._.mapKeys(action.payload, "id") };
    case ADD_CONTACT:
      return { ...state, [action.payload.id]: action.payload };
    case EDIT_CONTACT : 
      return {...state, [action.payload.id]: action.payload};
    //case DELETE_CONTACT: 
      //return _.omit(state, action.payload);
    default:
      return state;
  }
};
