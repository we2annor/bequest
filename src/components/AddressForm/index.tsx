import React, { useState, useEffect, ChangeEvent } from "react";
import Autosuggest from "react-autosuggest";
import { connect } from "react-redux";
import { addContact } from "../../actions";
import { countryNames } from "../../api/addressBook";
import { Address, AddressFormProps } from "../../actions/types";
import InputField from "../InputField";
import Error from "../Error";

interface IValidate {
  firstName: string;
  lastName: string;
  houseNo: string;
  postTown: string;
  country: string;
  postCode: string;
}

const AddressForm: React.FC<AddressFormProps> = ({
  selectedAddress,
  postcode,
  addContact,
}) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [houseNo, setHouseNo] = useState("");
  const [streetName, setStreetName] = useState("");
  const [locality, setLocality] = useState("");
  const [postTown, setPostTown] = useState("");
  const [country, setCountry] = useState("");
  const [countrySuggestions, setCountrySuggestions] = useState([]);
  const [suggestions, setSuggestions] = useState([]);
  const [postCode, setPostCode] = useState("");
  const [validationError, setValidationError] = useState({});

  useEffect(() => {
    const getSelectedAddress = () => {
      const address = Object.values(selectedAddress);
      setHouseNo(address[0].toString());
      setStreetName(address[1].toString());
      setLocality(address[2].toString());
      setPostTown(address[3].toString());
      setCountry(address[4].toString());
      setPostCode(postcode);
    };

    getSelectedAddress();
  }, [selectedAddress, postcode]);

  useEffect(() => {
    const fetchCountrySuggestions = async () => {
      if (country) {
        try {
          const response = await countryNames.get(`/name/${country}`);
          setCountrySuggestions(response.data);
        } catch (error) {
          console.log("country suggestion Error", error);
        }
      }
    };

    fetchCountrySuggestions();
  }, [country]);

  const handleOnChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    switch (name) {
      case "firstName":
        setFirstName(value);
        return;
      case "lastName":
        setLastName(value);
        return;
      case "houseNo":
        setHouseNo(value);
        return;
      case "streetName":
        setStreetName(value);
        return;
      case "locality":
        setLocality(value);
        return;
      case "postTown":
        setPostTown(value);
        return;
      case "country":
        console.log("val");
        setCountry(value);
        return;
      case "postCode":
        setPostCode(value);
        return;
    }
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (addressValidated()) {
      const formValues: Address = {
        firstName,
        lastName,
        houseNo,
        streetName,
        locality,
        postTown,
        country,
        postCode,
        id: 0,
      };
      addContact(formValues);
      clearAddressForm();
    }
  };

  const clearAddressForm = () => {
    setFirstName("");
    setLastName("");
    setHouseNo("");
    setStreetName("");
    setLocality("");
    setPostTown("");
    setCountry("");
    setPostCode("");
    setValidationError({});
  };

  const isEmpty = (obj: Object) => {
    for (const key in obj) {
      if (obj.hasOwnProperty(key)) return false;
    }
    return true;
  };

  const addressValidated = (): boolean => {
    const errors: IValidate = Object();
    if (!houseNo) {
      errors.houseNo = "House number or house name is required";
    }
    if (!postTown) {
      errors.postTown = "Town field is required";
    }
    if (!country) {
      errors.country = "Country field is required";
    }
    if (!postCode) {
      errors.postCode = "Post code field is required";
    }

    return isEmpty(errors);
  };

  const showError = () => {
    const errors = Object.values(validationError);
    const valErrors = errors.map((error, index) => (
      <li key={index} className='field-error__list'>
        <Error message={error as string} />
      </li>
    ));
    return valErrors;
  };

  const getSuggestions = (value: any) => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    return inputLength === 0
      ? []
      : countrySuggestions.filter(
          (suggestion: any) =>
            suggestion.name.toLowerCase().slice(0, inputLength) === inputValue
        );
  };

  const onSuggestionsFetchRequested = ({ value }: any) => {
    setSuggestions(getSuggestions(value));
  };

  const onSuggestionsClearedRequested = () => {
    setSuggestions([]);
  };

  const getSuggestionValue = (suggestion: any) => suggestion.name;
  const renderSuggestion = (suggestion: any) => (
    <div className='countryAutoSuggestion'>{suggestion.name}</div>
  );

  const onChange = (_event: any, { newValue }: any) => {
    setCountry(newValue);
  };

  const inputProps = {
    placeholder: "Country",
    autoComplete: "abcd",
    name: "country",
    id: "country",
    value: country,
    onChange: onChange,
  };

  return (
    <section className='contact-form'>
      <h2 className='u-margin-medium'>Add address manually</h2>
      <form className='contact' onSubmit={(e) => handleSubmit(e)}>
        {validationError && (
          <ul className='contact__field-errors'>{showError()}</ul>
        )}

        <div className='contact__full-name'>
          <InputField
            className='contact__full-name--first-name'
            label='First Name'
            id='firstName'
            value={firstName}
            autoComplete={"off"}
            placeholder='first name'
            onChange={(e) => handleOnChange(e)}
            required={false}
          />
          <InputField
            className='contact__full-name--last-name'
            label='Last Name'
            id='lastName'
            value={lastName}
            autoComplete={"off"}
            placeholder='last name'
            onChange={(e) => handleOnChange(e)}
            required={false}
          />
        </div>
        <div className='contact__address-lines'>
          <InputField
            className='contact__address-lines--house-no'
            label='House / Flat Number'
            id='houseNo'
            value={houseNo}
            autoComplete={"off"}
            placeholder='House / street name'
            onChange={(e) => handleOnChange(e)}
            required
          />
          <InputField
            className='contact__address-lines--street-name'
            label='Street Name'
            id='streetName'
            value={streetName}
            autoComplete='abcd'
            placeholder='Street name'
            onChange={(e) => handleOnChange(e)}
            required
          />
        </div>

        <div className='contact__area'>
          <InputField
            className='contact__area--locality'
            label='Locality'
            id='locality'
            value={locality}
            autoComplete='off'
            placeholder='Locality'
            onChange={(e) => handleOnChange(e)}
            required={false}
          />
          <InputField
            className='contact__area--post-town'
            label='Town'
            id='postTown'
            value={postTown}
            autoComplete='off'
            placeholder='Town'
            onChange={(e) => handleOnChange(e)}
            required
          />
        </div>

        <div className='contact__country'>
          <div className='input-field contact__country--autoSuggest-container'>
            <label htmlFor='country' aria-label='country'>
              Country
            </label>
            <Autosuggest
              suggestions={suggestions}
              onSuggestionsFetchRequested={onSuggestionsFetchRequested}
              onSuggestionsClearRequested={onSuggestionsClearedRequested}
              getSuggestionValue={getSuggestionValue}
              renderSuggestion={renderSuggestion}
              inputProps={inputProps}
            />
          </div>
          <InputField
            className='contact__country--post-code'
            label='Post code'
            id='postCode'
            value={postCode}
            autoComplete='off'
            placeholder='post code eg: Ha4 5gh'
            onChange={(e) => handleOnChange(e)}
            required
          />
        </div>
        <div className='contact__controls'>
          <button className='contact__controls--submit-btn btn' type='submit'>
            Submit
          </button>
          <button
            className='contact__controls--clear-btn btn'
            type='reset'
            onClick={clearAddressForm}
          >
            Clear
          </button>
        </div>
      </form>
    </section>
  );
};

export default connect(null, { addContact })(AddressForm);
