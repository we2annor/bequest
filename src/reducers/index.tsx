import {combineReducers} from 'redux';
import { addressReducer } from './addressReducer';

export const rootReducer = combineReducers({
    addressBook: addressReducer
})

export type RootState = ReturnType< typeof rootReducer>