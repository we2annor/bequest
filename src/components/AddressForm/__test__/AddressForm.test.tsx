import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { rootReducer } from "../../../reducers";
import AddressForm from "../index";
import { InitialAddressValues } from "../../../actions/types";

import renderer from "react-test-renderer";

afterEach(cleanup);

describe("AddressForm", () => {
  test("renders with default props", () => {
    const store = createStore(rootReducer, applyMiddleware(thunk));
    render(
      <Provider store={store}>
        <AddressForm
          selectedAddress={InitialAddressValues}
          postcode={"ha9 8rf"}
        />
      </Provider>
    );
  });

  test("Matches snapshot", () => {
    const store = createStore(rootReducer, applyMiddleware(thunk));
    const tree = renderer
      .create(
        <Provider store={store}>
          <AddressForm
            selectedAddress={InitialAddressValues}
            postcode={"ha9 8rf"}
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
