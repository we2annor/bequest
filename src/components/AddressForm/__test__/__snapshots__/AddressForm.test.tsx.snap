// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressForm Matches snapshot 1`] = `
<section
  className="contact-form"
  data-testid="contact-form"
>
  <h2
    className="u-margin-medium"
  >
    Add address manually
  </h2>
  <form
    className="contact"
    onSubmit={[Function]}
  >
    <ul
      className="contact__field-errors"
    />
    <div
      className="contact__full-name"
    >
      <div
        className="input-field contact__full-name--first-name"
      >
        <label
          aria-label="firstName"
          htmlFor="firstName"
        >
          First Name
        </label>
        <input
          autoComplete="off"
          id="firstName"
          name="firstName"
          onChange={[Function]}
          placeholder="first name"
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="input-field contact__full-name--last-name"
      >
        <label
          aria-label="lastName"
          htmlFor="lastName"
        >
          Last Name
        </label>
        <input
          autoComplete="abc"
          id="lastName"
          name="lastName"
          onChange={[Function]}
          placeholder="last name"
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="contact__address-lines"
    >
      <div
        className="input-field contact__address-lines--house-no"
      >
        <label
          aria-label="houseNo"
          htmlFor="houseNo"
        >
          House / Flat Number
          <span
            style={
              Object {
                "color": "red",
                "fontSize": "1.2rem",
              }
            }
          >
            required field *
          </span>
        </label>
        <input
          autoComplete="abcd"
          id="houseNo"
          name="houseNo"
          onChange={[Function]}
          placeholder="House / street name"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="input-field contact__address-lines--street-name"
      >
        <label
          aria-label="streetName"
          htmlFor="streetName"
        >
          Street Name
          <span
            style={
              Object {
                "color": "red",
                "fontSize": "1.2rem",
              }
            }
          >
            required field *
          </span>
        </label>
        <input
          autoComplete="abcd"
          id="streetName"
          name="streetName"
          onChange={[Function]}
          placeholder="Street name"
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="contact__area"
    >
      <div
        className="input-field contact__area--locality"
      >
        <label
          aria-label="locality"
          htmlFor="locality"
        >
          Locality
        </label>
        <input
          autoComplete="off"
          id="locality"
          name="locality"
          onChange={[Function]}
          placeholder="Locality"
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="input-field contact__area--post-town"
      >
        <label
          aria-label="postTown"
          htmlFor="postTown"
        >
          City
          <span
            style={
              Object {
                "color": "red",
                "fontSize": "1.2rem",
              }
            }
          >
            required field *
          </span>
        </label>
        <input
          autoComplete="off"
          id="postTown"
          name="postTown"
          onChange={[Function]}
          placeholder="Town"
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="contact__country"
    >
      <div
        className="input-field contact__country--autoSuggest-container"
      >
        <label
          aria-label="country"
          htmlFor="country"
        >
          Country
        </label>
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-owns="react-autowhatever-1"
          className="react-autosuggest__container"
          role="combobox"
        >
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls="react-autowhatever-1"
            autoComplete="abcd"
            className="react-autosuggest__input"
            id="country"
            name="country"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Country"
            type="text"
            value=""
          />
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-1"
            role="listbox"
          />
        </div>
      </div>
      <div
        className="input-field contact__country--post-code"
      >
        <label
          aria-label="postCode"
          htmlFor="postCode"
        >
          Post code
          <span
            style={
              Object {
                "color": "red",
                "fontSize": "1.2rem",
              }
            }
          >
            required field *
          </span>
        </label>
        <input
          autoComplete="off"
          id="postCode"
          name="postCode"
          onChange={[Function]}
          placeholder="post code eg: Ha4 5gh"
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="contact__controls"
    >
      <button
        className="contact__controls--submit-btn btn"
        type="submit"
      >
        Submit
      </button>
      <button
        className="contact__controls--clear-btn btn"
        onClick={[Function]}
        type="reset"
      >
        Clear
      </button>
    </div>
  </form>
</section>
`;
